### PR TITLE
Add function to get standard deviation (CBA_fnc_standardDeviation)

### PR DESCRIPTION
**Pull request by Kexanone**
**When merged this pull request will:**
- Implement `CBA_fnc_standardDeviation`

### CODE REVIEW

File: addons/arrays/fnc_stdDev.sqf [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    [[1,4,16,4,1], 1] call CBA_fnc_stdDev;
+    (end)
+
+Author:
+    Kex
+---------------------------------------------------------------------------- */
+
+params [["_array", [], [[]]], ["_ddof", 0, [0]]];
+
+private _N = (count _array) - _ddof;
```
Comment (commy2) [inline addons/arrays/fnc_stdDev.sqf:32]: camel case (`_n`)
superfluous parentheses
`_array` -> `_numbers`
Not a fan of abreviated variable names.

File: addons/arrays/fnc_stdDev.sqf [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+if (_N <= 0) exitWith {0};
+
+private _sqrResSum = 0;
```
Comment (commy2) [inline addons/arrays/fnc_stdDev.sqf:35]: What does that mean? "squared residual sum"? If anything it should be `_resSumSqrs`

File: addons/arrays/fnc_stdDev.sqf [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+private _mean = _array call BIS_fnc_arithmeticMean;
```
Comment (commy2) [inline addons/arrays/fnc_stdDev.sqf:36]: Somehow I don't like relying on BIS functions for trivial stuff like this.

**Comment by commy2** [comment]
If I can call
```sqf
private _draw = [1, 2, 3];
_draw call BIS_fnc_arithmeticMean
```
like this, why can't I call
```
private _draw = [1, 2, 3];
_draw call CBA_fnc_stdDev
```
like this?
Why ist one written out full name and the other abbreviated?

Not an expert on this, but is there any meaning to calling this with anything else other than 0 or 1 as second argument?

**Comment by commy2** [comment]
This should have a unit test tbh.

**Comment by Kexanone** [comment]
Should have made it as a draft 😅

**Comment by Kexanone** [comment]
> Not an expert on this, but is there any meaning to calling this with anything else other than 0 or 1 as second argument?

I took `numpy.std` as inspiration, where they also use ddof as optional parameter to handle the two cases.

**Comment by commy2** [comment]
>Should have made it as a draft 😅

No worries, just want to get it right.
Also, never understood the point of drafts. Just label your pr as such if it is wip.

**Comment by veteran29** [comment]
> Also, never understood the point of drafts. Just label your pr as such if it is wip.

Drafts are useful if you have auto-requests for reviews for PRs based on code ownership. Which is not a case in this repo.

File: addons/arrays/fnc_standardDeviation.sqf [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    Kex
+---------------------------------------------------------------------------- */
+
+params [["_array", [], [[]]], ["_ddof", 0, [0]]];
+
+private _count = count _array;
+if (_count <= _ddof) exitWith {0};
+
+private _mean = 0;
+{_mean = _mean + _x} forEach _array;
```
Comment (commy2) [inline addons/arrays/fnc_standardDeviation.sqf:36]: old was better

File: addons/arrays/fnc_standardDeviation.sqf [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    [[1,4,16,4,1]] call CBA_fnc_standardDeviation;
+    // returns roughly 6.221
+    [[1,4,16,4,1], 1] call CBA_fnc_standardDeviation;
+    (end)
+
+Author:
+    Kex
+---------------------------------------------------------------------------- */
+
+params [["_array", [], [[]]], ["_ddof", 0, [0]]];
```
Comment (commy2) [inline addons/arrays/fnc_standardDeviation.sqf:30]: _numbers

File: addons/arrays/fnc_standardDeviation.sqf (hunk 4 above)
Comment (Kexanone) [reply addons/arrays/fnc_standardDeviation.sqf:36]: So stick to BIS function?

**Comment by Kexanone** [comment]
Side note: Test for `arrays` component fails because of other functions, but my part is fine.

File: addons/arrays/test_standardDeviation.sqf [hunk 6]
```diff
@@ -0,0 +1,32 @@
+// ----------------------------------------------------------------------------
+#define DEBUG_MODE_FULL
+#define DEBUG_SYNCHRONOUS
+#include "script_component.hpp"
+
+SCRIPT(test_standardDeviation);
+
+// ----------------------------------------------------------------------------
+
+private ["_original", "_expected", "_result", "_fn"];
```
Comment (commy2) [inline addons/arrays/test_standardDeviation.sqf:10]: _original is never used, so remove.
Otherwise it all lgtm

**Comment by Kexanone** [comment]
```
17:25:40 [CBA] (arrays) LOG: Testing CBA_fnc_findMax
17:25:40 [CBA] (arrays) Test OK: (CBA_fnc_findMax is defined) x\cba\addons\arrays\test_findMax.sqf:15
17:25:40 [CBA] (arrays) Test OK: (_result isEqualTo _expected) x\cba\addons\arrays\test_findMax.sqf:20
17:25:40 [CBA] (arrays) Test OK: (_result isEqualTo _expected) x\cba\addons\arrays\test_findMax.sqf:25
17:25:40 [CBA] (arrays) Test OK: (_result isEqualTo _expected) x\cba\addons\arrays\test_findMax.sqf:30
17:25:40 [CBA] (arrays) Test OK: (_result isEqualTo _expected) x\cba\addons\arrays\test_findMax.sqf:35
17:25:40 [CBA] (arrays) Test OK: (_result isEqualTo _expected) x\cba\addons\arrays\test_findMax.sqf:40
17:25:40 [CBA] (arrays) Test OK: (_result isEqualTo _expected) x\cba\addons\arrays\test_findMax.sqf:45
17:25:40 [CBA] (arrays) Test OK: (isNil "_result") x\cba\addons\arrays\test_findMax.sqf:49
17:25:40 [CBA] (arrays) Test OK: (isNil "_result") x\cba\addons\arrays\test_findMax.sqf:53
17:25:40 [CBA] (arrays) Test OK: (isNil "_result") x\cba\addons\arrays\test_findMax.sqf:57
17:25:40 [CBA] (arrays) Test OK: (isNil "_result") x\cba\addons\arrays\test_findMax.sqf:61
17:25:40 [CBA] (arrays) LOG: Testing CBA_fnc_standardDeviation
17:25:40 [CBA] (arrays) Test OK: (CBA_fnc_standardDeviation is defined) x\cba\addons\arrays\test_standardDeviation.sqf:15
17:25:40 [CBA] (arrays) Test OK: (_result isEqualTo _expected) x\cba\addons\arrays\test_standardDeviation.sqf:20
17:25:40 [CBA] (arrays) Test OK: (_result isEqualTo _expected) x\cba\addons\arrays\test_standardDeviation.sqf:25
17:25:40 [CBA] (arrays) Test OK: (_result isEqualTo _expected) x\cba\addons\arrays\test_standardDeviation.sqf:30
17:25:40 [CBA] (arrays) LOG: === Testing Arrays ===
17:25:40 [CBA] (arrays) LOG: Testing CBA_fnc_filter
17:25:40 [CBA] (arrays) Test OK: (CBA_fnc_filter is defined) x\cba\addons\arrays\test_filter.sqf:14
17:25:40 [CBA] (arrays) Test OK: (str _result == str _expected) x\cba\addons\arrays\test_filter.sqf:20
17:25:40 [CBA] (arrays) Test OK: (str _original == str _expected) x\cba\addons\arrays\test_filter.sqf:21
17:25:40 [CBA] (arrays) Test OK: (str _result == str _expected) x\cba\addons\arrays\test_filter.sqf:26
17:25:40 [CBA] (arrays) Test OK: (str _original != str _expected) x\cba\addons\arrays\test_filter.sqf:27
17:25:40 [CBA] (arrays) Test OK: (str _result == str _expected) x\cba\addons\arrays\test_filter.sqf:32
17:25:40 [CBA] (arrays) Test OK: (str _original == str _expected) x\cba\addons\arrays\test_filter.sqf:38
17:25:40 [CBA] (arrays) Test OK: (str _result == str _expected) x\cba\addons\arrays\test_filter.sqf:39
17:25:40 [CBA] (arrays) Test OK: (str _original == str _expected) x\cba\addons\arrays\test_filter.sqf:44
17:25:40 [CBA] (arrays) Test OK: (str _result == str _expected) x\cba\addons\arrays\test_filter.sqf:45
17:25:40 [CBA] (arrays) Test OK: (str _original == str _expected) x\cba\addons\arrays\test_filter.sqf:50
17:25:40 [CBA] (arrays) Test OK: (str _result == str _expected) x\cba\addons\arrays\test_filter.sqf:51
17:25:40 [CBA] (arrays) LOG: Testing CBA_fnc_inject
17:25:40 [CBA] (arrays) Test OK: (CBA_fnc_inject is defined) x\cba\addons\arrays\test_inject.sqf:14
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_inject.sqf:18
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_inject.sqf:22
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_inject.sqf:26
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_inject.sqf:30
17:25:40 [CBA] (arrays) LOG: Testing CBA_fnc_join
17:25:40 [CBA] (arrays) Test OK: (CBA_fnc_join is defined) x\cba\addons\arrays\test_join.sqf:14
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_join.sqf:18
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_join.sqf:22
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_join.sqf:26
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_join.sqf:30
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_join.sqf:34
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_join.sqf:38
17:25:40 [CBA] (arrays) LOG: Testing CBA_fnc_shuffle
17:25:40 [CBA] (arrays) Test OK: (CBA_fnc_shuffle is defined) x\cba\addons\arrays\test_shuffle.sqf:14
17:25:40 [CBA] (arrays) Test OK: (count _result == count _original) x\cba\addons\arrays\test_shuffle.sqf:18
17:25:40 [CBA] (arrays) Test OK: (_x in _original) x\cba\addons\arrays\test_shuffle.sqf:21
17:25:40 [CBA] (arrays) Test OK: (_x in _original) x\cba\addons\arrays\test_shuffle.sqf:21
17:25:40 [CBA] (arrays) Test OK: (_x in _original) x\cba\addons\arrays\test_shuffle.sqf:21
17:25:40 [CBA] (arrays) Test OK: (count _result == count _original) x\cba\addons\arrays\test_shuffle.sqf:26
17:25:40 [CBA] (arrays) Test OK: ((_original select 0) == 100) x\cba\addons\arrays\test_shuffle.sqf:31
17:25:40 [CBA] (arrays) LOG: Testing CBA_fnc_findNil
17:25:40 [CBA] (arrays) Test OK: (CBA_fnc_findNil is defined) x\cba\addons\arrays\test_findNil.sqf:15
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findNil.sqf:20
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findNil.sqf:25
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findNil.sqf:30
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findNil.sqf:35
17:25:40 [CBA] (arrays) LOG: Testing CBA_fnc_findNull
17:25:40 [CBA] (arrays) Test OK: (CBA_fnc_findNull is defined) x\cba\addons\arrays\test_findNull.sqf:15
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findNull.sqf:20
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findNull.sqf:25
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findNull.sqf:30
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findNull.sqf:35
17:25:40 [CBA] (arrays) LOG: Testing CBA_fnc_findTypeName
17:25:40 [CBA] (arrays) Test OK: (CBA_fnc_findTypeName is defined) x\cba\addons\arrays\test_findTypeName.sqf:15
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeName.sqf:22
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeName.sqf:26
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeName.sqf:30
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeName.sqf:34
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeName.sqf:38
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeName.sqf:42
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeName.sqf:46
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeName.sqf:50
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeName.sqf:54
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeName.sqf:58
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeName.sqf:62
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeName.sqf:66
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeName.sqf:70
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeName.sqf:74
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeName.sqf:78
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeName.sqf:82
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeName.sqf:87
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeName.sqf:92
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeName.sqf:97
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeName.sqf:102
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeName.sqf:107
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeName.sqf:112
17:25:40 [CBA] (arrays) LOG: Testing CBA_fnc_findTypeOf
17:25:40 [CBA] (arrays) Test OK: (CBA_fnc_findTypeOf is defined) x\cba\addons\arrays\test_findTypeOf.sqf:14
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeOf.sqf:19
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeOf.sqf:24
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeOf.sqf:29
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeOf.sqf:34
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeOf.sqf:39
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeOf.sqf:44
17:25:40 [CBA] (arrays) Test OK: (_result == _expected) x\cba\addons\arrays\test_findTypeOf.sqf:49
17:25:40 [CBA] (arrays) LOG: Testing CBA_fnc_findMin
17:25:40 [CBA] (arrays) Test OK: (CBA_fnc_findMin is defined) x\cba\addons\arrays\test_findMin.sqf:14
17:25:40 [CBA] (arrays) Test OK: (_result isEqualTo _expected) x\cba\addons\arrays\test_findMin.sqf:19
17:25:40 [CBA] (arrays) Test OK: (_result isEqualTo _expected) x\cba\addons\arrays\test_findMin.sqf:24
17:25:40 [CBA] (arrays) Test OK: (_result isEqualTo _expected) x\cba\addons\arrays\test_findMin.sqf:29
17:25:40 [CBA] (arrays) Test OK: (_result isEqualTo _expected) x\cba\addons\arrays\test_findMin.sqf:34
17:25:40 [CBA] (arrays) Test OK: (_result isEqualTo _expected) x\cba\addons\arrays\test_findMin.sqf:39
17:25:40 [CBA] (arrays) Test OK: (_result isEqualTo _expected) x\cba\addons\arrays\test_findMin.sqf:44
17:25:40 [CBA] (arrays) Test OK: (isNil "_result") x\cba\addons\arrays\test_findMin.sqf:48
17:25:40 [CBA] (arrays) Test OK: (isNil "_result") x\cba\addons\arrays\test_findMin.sqf:52
17:25:40 [CBA] (arrays) Test OK: (isNil "_result") x\cba\addons\arrays\test_findMin.sqf:56
17:25:40 [CBA] (arrays) Test OK: (isNil "_result") x\cba\addons\arrays\test_findMin.sqf:60
17:25:40 [CBA] (arrays) LOG: Testing CBA_fnc_insert
17:25:40 [CBA] (arrays) Test OK: (CBA_fnc_insert is defined) x\cba\addons\arrays\test_insert.sqf:14
17:25:40 [CBA] (arrays) Test OK: (_result isEqualTo _expected) x\cba\addons\arrays\test_insert.sqf:20
17:25:40 [CBA] (arrays) Test OK: (_result isEqualTo _expected) x\cba\addons\arrays\test_insert.sqf:26
```

File: addons/arrays/fnc_standardDeviation.sqf (hunk 4 above)
Comment (Kexanone) [reply addons/arrays/fnc_standardDeviation.sqf]: I agreed more with your earlier statement that it's not necessary to rely on a BIS function for trivial stuff. Hence, I would leave it as it is.

**Comment by commy2** [comment]
Really?! Remove tests in the same PR you add features?

File: addons/arrays/fnc_standardDeviation.sqf [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+if (_count <= _ddof) exitWith {0};
+
+private _mean = 0;
+{_mean = _mean + _x} count _numbers;
+_mean = _mean / _count;
+
+private _resSumSqrs = 0;
+{
+    _resSumSqrs = _resSumSqrs + (_x - _mean)^2;
+} count _numbers;
```
Comment (commy2) [inline addons/arrays/fnc_standardDeviation.sqf:42]: `forEach`

File: addons/arrays/fnc_standardDeviation.sqf [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    Kex
+---------------------------------------------------------------------------- */
+
+params [["_numbers", [], [[]]], ["_ddof", 0, [0]]];
+
+private _count = count _numbers;
+if (_count <= _ddof) exitWith {0};
+
+private _mean = 0;
+{_mean = _mean + _x} count _numbers;
```
Comment (commy2) [inline addons/arrays/fnc_standardDeviation.sqf:36]: `forEach`

File: addons/arrays/fnc_standardDeviation.sqf (hunk 7 above)
Comment (Kexanone) [reply addons/arrays/fnc_standardDeviation.sqf:42]: I actually checked `BIS_fnc_arithmeticMean` and they also prefer `count` over `forEach`. KK has apparently marked it as an optimization.

**Comment by Kexanone** [comment]
> Really?! Remove tests in the same PR you add features?

How am I supposed to write and validate unit tests when the component is completely broken?

**Comment by Kexanone** [comment]
I removed it, but at least we know now that it works with the patch.

File: addons/arrays/fnc_standardDeviation.sqf (hunk 7 above)
Comment (commy2) [reply addons/arrays/fnc_standardDeviation.sqf:42]: Use `forEach`.

**Comment by commy2** [comment]
```sqf
execVM "\x\cba\addons\arrays\test_standardDeviation.sqf"
```

```
21:37:08 [CBA] (arrays) LOG: Testing CBA_fnc_standardDeviation
21:37:08 [CBA] (arrays) Test OK: (CBA_fnc_standardDeviation is defined) x\cba\addons\arrays\test_standardDeviation.sqf:15
21:37:08 [CBA] (arrays) Test OK: (_result isEqualTo _expected) x\cba\addons\arrays\test_standardDeviation.sqf:20
21:37:08 [CBA] (arrays) Test OK: (_result isEqualTo _expected) x\cba\addons\arrays\test_standardDeviation.sqf:25
21:37:08 [CBA] (arrays) Test OK: (_result isEqualTo _expected) x\cba\addons\arrays\test_standardDeviation.sqf:30
```

**Review by commy2 — APPROVED** [review]

**Review by jonpas — APPROVED** [review]

**Pull request merged**